src/reader.cpp

```diff
@@ -1288,11 +1288,21 @@ static history_pager_result_t history_pager_search(const std::shared_ptr<history
     completion_list_t completions;
     history_search_t search{history, search_string, history_search_type_t::contains,
                             smartcase_flags(search_string), history_index};
-    while (completions.size() < page_size && search.go_to_next_match(direction)) {
+    bool next_match_found = search.go_to_next_match(direction);
+    if (!next_match_found) {
+        // If there were no matches, try again with subsequence search
+        history_search_t search{history, search_string, history_search_type_t::contains_subsequence,
+                                smartcase_flags(search_string), history_index};
+        next_match_found = search.go_to_next_match(direction);
+        // TODO next_match_found is 1 even if there are no matches
```

Comment:
`go_to_next_match` seems to return `true` even when there are no matches. I don't know else to determine whether there were no matches.

Reply:
With this small tweak, it seems to work fine for me.
The problem was that the fallback search goes out of scope, so it would always use the substring search.

```diff
diff --git a/src/reader.cpp b/src/reader.cpp
index a3b54126d..174b48cd5 100644
--- a/src/reader.cpp
+++ b/src/reader.cpp
@@ -1327,10 +1327,10 @@ static history_pager_result_t history_pager_search(const std::shared_ptr<history
     bool next_match_found = search.go_to_next_match(direction);
     if (!next_match_found) {
         // If there were no matches, try again with subsequence search
-        history_search_t search{history, search_string, history_search_type_t::contains_subsequence,
-                                smartcase_flags(search_string), history_index};
+        search =
+            history_search_t{history, search_string, history_search_type_t::contains_subsequence,
+                             smartcase_flags(search_string), history_index};
         next_match_found = search.go_to_next_match(direction);
-        // TODO next_match_found is 1 even if there are no matches
     }
     while (completions.size() < page_size && next_match_found) {
         const history_item_t &item = search.current_item();
```